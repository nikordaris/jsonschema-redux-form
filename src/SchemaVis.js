// @flow
import React, { Component } from 'react';
import { get, set, merge, omit, isEmpty, isEqual } from 'lodash';

import {
  DEFAULT_PREFIX,
  isDisabled,
  getComponent,
  getStyle,
  getPrefix
} from './selectors';

import { getComparator, getDefaultComponent, isRequired } from './utils';

export default class SchemaVis extends Component {
  static defaultProps = {
    styles: {},
    prefix: DEFAULT_PREFIX,
    tag: 'div',
    id: 'schemaVis'
  };

  props: {
    schema: SchemaType,
    definitions: { [string]: SchemaType },
    id: string,
    prefix: string,
    namespace?: string,
    styles: SchemaVisStylesType,
    components: { [string]: React.Element<*> | string },
    defaultComponents: any,
    componentProps: { [string]: { styles: { [string]: any } } },
    tag: string
  };

  state: {
    definitions: { [string]: SchemaType }
  };

  componentWillMount() {
    const {
      schema: { definitions: schemaDefinitions } = {},
      definitions: propDefinitions
    } = this.props;
    this.setDefinitions(schemaDefinitions, propDefinitions);
  }

  componentWillReceiveProps(nextProps: any) {
    const {
      definitions: nextPropDefinitions,
      schema: { definitions: nextSchemaDefinitions } = {}
    } = nextProps;
    const {
      definitions: propDefinitions,
      schema: { definitions: schemaDefinitions } = {}
    } = this.props;
    if (
      !isEqual(nextPropDefinitions, propDefinitions) ||
      !isEqual(nextSchemaDefinitions, schemaDefinitions)
    ) {
      this.setDefinitions(nextSchemaDefinitions, nextPropDefinitions);
    }
  }

  setDefinitions(
    schemaDefinitions: { [string]: SchemaType } = {},
    propDefinitions: { [string]: SchemaType } = {}
  ) {
    this.setState({
      ...this.state,
      definitions: merge({}, schemaDefinitions, propDefinitions)
    });
  }

  setSchemaDefMetadata(schema: SchemaType, metadata: any) {
    const { prefix } = this.props;
    if (isEmpty(prefix)) {
      return merge({}, schema, metadata);
    }

    return set(schema, prefix, metadata);
  }

  getSchema(schema: SchemaType) {
    const { definitions } = this.state;
    const { prefix } = this.props;
    const ref = get(schema, '$ref');
    if (ref) {
      // expects #/definitions/[PATH/TO/DEF]
      const refDot = ref.slice(14).replace('/', '.');
      const refSchema = get(definitions, refDot);
      return (
        refSchema &&
        this.setSchemaDefMetadata(
          refSchema,
          merge(
            {},
            getPrefix(omit(schema, ['$ref']), prefix),
            getPrefix(refSchema, prefix)
          )
        )
      );
    }
    return schema;
  }

  renderChildren = (schema: SchemaType, namespace?: string) => {
    const { prefix } = this.props;
    const children = schema.type === 'object' && schema.properties
      ? schema.properties
      : {};
    return Object.keys(children)
      .filter(prop => !isDisabled(children[prop], prefix))
      .sort(getComparator(children, prefix))
      .map((prop, idx) =>
        this.renderSchema(
          children[prop],
          idx,
          namespace ? `${namespace}.${prop}` : prop,
          isRequired(schema, prop)
        )
      );
  };

  renderSchema = (
    schema: SchemaType,
    idx: number | string,
    name?: string,
    required: boolean = false
  ) => {
    const {
      styles: { component: componentStyles = {} },
      components,
      componentProps,
      prefix,
      tag: Tag,
      defaultComponents
    } = this.props;
    const schemaDef = this.getSchema(schema);
    const schemaStyle = getStyle(schemaDef, prefix, {});
    const component = getComponent(schemaDef, prefix);
    const rest = omit(this.props, [
      'schema',
      'prefix',
      'styles',
      'namespace',
      'components',
      'componentProps',
      'defaultComponents',
      'tag'
    ]);

    const ComponentVis = get(
      components,
      component,
      getDefaultComponent(defaultComponents, schemaDef)
    );

    if (ComponentVis) {
      const { styles: componentPropStyles = {}, ...componentProp } = get(
        componentProps,
        component,
        {}
      );

      const componentAttributes = {
        styles: merge({}, componentStyles, componentPropStyles, schemaStyle),
        key: idx,
        name,
        required,
        schemaVis: {
          prefix,
          schema: schemaDef,
          components,
          componentProps,
          defaultComponents
        },
        ...componentProp,
        ...rest
      };

      if (React.isValidElement(ComponentVis)) {
        return React.cloneElement(ComponentVis, componentAttributes);
      }

      return React.createElement(ComponentVis, componentAttributes);
    } else if (!isEmpty(get(schemaDef, 'properties', {}))) {
      return React.createElement(
        Tag,
        { key: idx, ...rest },
        this.renderChildren(schemaDef, name)
      );
    }

    return undefined;
  };

  render() {
    const { schema, namespace, id } = this.props;

    return this.renderSchema(schema, id, namespace) || <div />;
  }
}
