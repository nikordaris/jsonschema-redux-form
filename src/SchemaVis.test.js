import React from 'react';
import renderer from 'react-test-renderer';

import SchemaVis from './index.js';

const schema = {
  title: 'Test',
  type: 'object',
  required: ['foo', 'foobar'],
  properties: {
    foo: {
      title: 'Foo',
      type: 'number',
      meta: {
        vis: {
          editable: true,
          component: 'NumberInputField'
        }
      }
    },
    bar: {
      title: 'Bar',
      type: 'string',
      required: true,
      meta: {
        vis: {
          ordinal: 1,
          editable: true,
          component: 'InputField'
        }
      }
    },
    foobar: {
      title: 'Foobar',
      description: 'Foo is bar',
      type: 'string',
      meta: {
        vis: {
          editable: true,
          component: 'InputField'
        }
      }
    },
    foobar2: {
      title: 'Foobar2',
      description: 'Foo is bar',
      type: 'string',
      meta: {
        vis: {
          ordinal: 3,
          editable: true
        }
      }
    },
    blah: {
      title: 'Blah',
      type: 'string',
      meta: {
        vis: {
          editable: true,
          ordinal: 1
        }
      }
    },
    blah2: {
      title: 'Blah2',
      type: 'string'
    }
  }
};

const components = {
  NumberInputField: <input type="number" />,
  InputField: 'input'
};

describe('Render Schema', () => {
  it('should render', () => {
    const tree = renderer
      .create(<SchemaVis schema={schema} components={components} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without required fields', () => {
    const tree = renderer
      .create(
        <SchemaVis
          schema={{ ...schema, required: undefined }}
          components={components}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render single field schema', () => {
    const tree = renderer
      .create(
        <SchemaVis
          namespace="bar"
          schema={{
            title: 'Foo',
            type: 'number',
            meta: {
              vis: {
                ordinal: 0,
                editable: true,
                component: 'NumberInputField'
              }
            }
          }}
          components={components}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without properties', () => {
    const tree = renderer
      .create(
        <SchemaVis
          schema={{
            title: 'Test',
            type: 'object'
          }}
          components={components}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without name and with namespace', () => {
    const tree = renderer
      .create(
        <SchemaVis schema={schema} namespace="foo" components={components} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
