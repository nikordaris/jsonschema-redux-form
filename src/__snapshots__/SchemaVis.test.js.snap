// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Schema should render 1`] = `
<div
  id="schemaVis"
>
  <input
    id="schemaVis"
    name="foo"
    required={true}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "NumberInputField",
              "editable": true,
            },
          },
          "title": "Foo",
          "type": "number",
        },
      }
    }
    styles={Object {}}
    type="number"
  />
  <input
    id="schemaVis"
    name="foobar"
    required={true}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "description": "Foo is bar",
          "meta": Object {
            "vis": Object {
              "component": "EmailInputField",
              "editable": true,
            },
          },
          "title": "Foobar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
    type="email"
  />
  <input
    id="schemaVis"
    name="bar"
    required={false}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "InputField",
              "editable": true,
              "ordinal": 1,
            },
          },
          "required": true,
          "title": "Bar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
  />
</div>
`;

exports[`Render Schema should render single field schema 1`] = `
<input
  id="schemaVis"
  name="bar"
  required={false}
  schemaVis={
    Object {
      "componentProps": undefined,
      "components": Object {
        "EmailInputField": [Function],
        "InputField": "input",
        "NumberInputField": <input
          type="number"
    />,
      },
      "defaultComponents": undefined,
      "prefix": "meta.vis",
      "schema": Object {
        "meta": Object {
          "vis": Object {
            "component": "NumberInputField",
            "editable": true,
            "ordinal": 0,
          },
        },
        "title": "Foo",
        "type": "number",
      },
    }
  }
  styles={Object {}}
  type="number"
/>
`;

exports[`Render Schema should render without name and with namespace 1`] = `
<div
  id="schemaVis"
>
  <input
    id="schemaVis"
    name="foo.foo"
    required={true}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "NumberInputField",
              "editable": true,
            },
          },
          "title": "Foo",
          "type": "number",
        },
      }
    }
    styles={Object {}}
    type="number"
  />
  <input
    id="schemaVis"
    name="foo.foobar"
    required={true}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "description": "Foo is bar",
          "meta": Object {
            "vis": Object {
              "component": "EmailInputField",
              "editable": true,
            },
          },
          "title": "Foobar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
    type="email"
  />
  <input
    id="schemaVis"
    name="foo.bar"
    required={false}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "InputField",
              "editable": true,
              "ordinal": 1,
            },
          },
          "required": true,
          "title": "Bar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
  />
</div>
`;

exports[`Render Schema should render without properties 1`] = `<div />`;

exports[`Render Schema should render without required fields 1`] = `
<div
  id="schemaVis"
>
  <input
    id="schemaVis"
    name="foo"
    required={false}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "NumberInputField",
              "editable": true,
            },
          },
          "title": "Foo",
          "type": "number",
        },
      }
    }
    styles={Object {}}
    type="number"
  />
  <input
    id="schemaVis"
    name="foobar"
    required={false}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "description": "Foo is bar",
          "meta": Object {
            "vis": Object {
              "component": "EmailInputField",
              "editable": true,
            },
          },
          "title": "Foobar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
    type="email"
  />
  <input
    id="schemaVis"
    name="bar"
    required={false}
    schemaVis={
      Object {
        "componentProps": undefined,
        "components": Object {
          "EmailInputField": [Function],
          "InputField": "input",
          "NumberInputField": <input
            type="number"
      />,
        },
        "defaultComponents": undefined,
        "prefix": "meta.vis",
        "schema": Object {
          "meta": Object {
            "vis": Object {
              "component": "InputField",
              "editable": true,
              "ordinal": 1,
            },
          },
          "required": true,
          "title": "Bar",
          "type": "string",
        },
      }
    }
    styles={Object {}}
  />
</div>
`;
